src/dawn/CodeGen/Cuda/CudaCodeGen.cpp

```diff
@@ -0,0 +1,1139 @@
+  cudaKernel.addComment("Regions (a,h,e) and (c,i,g) are executed by two specialized warp");
+
+  // jboundary_limit determines the number of warps required to execute (b,d,f)");
+  int jboundary_limit = (int)nty + +maxExtents[1].Plus - maxExtents[1].Minus;
+  int iminus_limit = jboundary_limit + (maxExtents[0].Minus < 0 ? 1 : 0);
+  int iplus_limit = iminus_limit + (maxExtents[0].Plus > 0 ? 1 : 0);
+
+  cudaKernel.addStatement("int iblock = " + std::to_string(maxExtents[0].Minus) + " - 1");
+  cudaKernel.addStatement("int jblock = " + std::to_string(maxExtents[1].Minus) + " - 1");
+  cudaKernel.addBlockStatement("if(threadIdx.y < +" + std::to_string(jboundary_limit) + ")", [&]() {
+    cudaKernel.addStatement("iblock = threadIdx.x");
+    cudaKernel.addStatement("jblock = (int)threadIdx.y + " + std::to_string(maxExtents[1].Minus));
+  });
+  if(maxExtents[0].Minus < 0) {
+    cudaKernel.addBlockStatement(
+        "else if(threadIdx.y < +" + std::to_string(iminus_limit) + ")", [&]() {
+          int paddedBoundary_ = paddedBoundary(maxExtents[0].Minus);
+
+          // we dedicate one warp to execute regions (a,h,e), so here we make sure we have enough
+          //  threads
+          DAWN_ASSERT_MSG((jboundary_limit * paddedBoundary_ <= blockSize[0]),
+                          "not enought cuda threads");
+
+          cudaKernel.addStatement("iblock = -" + std::to_string(paddedBoundary_) +
+                                  " + (int)threadIdx.x % " + std::to_string(paddedBoundary_));
+          cudaKernel.addStatement("jblock = (int)threadIdx.x / " + std::to_string(paddedBoundary_) +
+                                  "+" + std::to_string(maxExtents[1].Minus));
+        });
+  }
+  if(maxExtents[0].Plus > 0) {
+    cudaKernel.addBlockStatement(
+        "else if(threadIdx.y < " + std::to_string(iplus_limit) + ")", [&]() {
+          int paddedBoundary_ = paddedBoundary(maxExtents[0].Plus);
+          // we dedicate one warp to execute regions (c,i,g), so here we make sure we have enough
+          //    threads
+          // we dedicate one warp to execute regions (a,h,e), so here we make sure we have enough
+          //  threads
+          DAWN_ASSERT_MSG((jboundary_limit * paddedBoundary_ <= blockSize[0]),
+                          "not enought cuda threads");
+
+          cudaKernel.addStatement("iblock = threadIdx.x % " + std::to_string(paddedBoundary_) +
+                                  " + " + std::to_string(ntx));
+          cudaKernel.addStatement("jblock = (int)threadIdx.x / " + std::to_string(paddedBoundary_) +
+                                  "+" + std::to_string(maxExtents[1].Minus));
+        });
+  }
+
+  std::unordered_map<int, Array3i> fieldIndexMap;
+  std::unordered_map<std::string, Array3i> indexIterators;
+
+  for(auto field : nonTempFields) {
+    Array3i dims{-1, -1, -1};
+    // TODO this is a hack, we need to have dimensions also at ms level
```

Comment:
is this fixed or should we wait on this?

Reply:
fixed